[MM-19533] Add toasts feature

sudheerDev opened the pull request:
#### Summary
This PR adds toasts feature for webapp based on the spec https://docs.google.com/document/d/1lANE8O3jVn4yjHNBbe4tMtXVcl3BYcRLWS0IViQ2Vds/edit

Toasts are of two types 
1. Unread toasts which appears when entering a channel with unreads
<img width="1665" alt="Screenshot 2020-01-08 at 2 20 59 AM" src="https://user-images.githubusercontent.com/4973621/71922474-8baa5f80-31bd-11ea-919d-ff3af31457d4.png">

2. New messages toasts which appears when new posts arrive and user is not at the bottom of the channel
<img width="1684" alt="Screenshot 2020-01-08 at 2 19 36 AM" src="https://user-images.githubusercontent.com/4973621/71922493-9664f480-31bd-11ea-98e2-4054230cb7dc.png">

3. Similar toast like 1. When users marks a posts as unread

This is a new PR made from https://github.com/mattermost/mattermost-webapp/pull/4132. Considering the PM review done based on that.

#### Ticket Link
[MM-19537](https://mattermost.atlassian.net/browse/MM-19537)
[MM-19533](https://mattermost.atlassian.net/browse/MM-19533)

#### Related Pull Requests
requires redux PR for a new selector: https://github.com/mattermost/mattermost-redux/pull/967

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

jgilliam17 commented:
Just noting here, as this is something I observed while testing
When both toasts (channel sidebar and new toast) are both shown side by side, the difference in design is noticeable. Is this ok?
<img width="964" alt="Screen Shot 2020-01-07 at 6 06 21 PM" src="https://user-images.githubusercontent.com/52937121/71937595-d8207b80-317a-11ea-9349-cbdd4448995c.png">
<img width="968" alt="Screen Shot 2020-01-07 at 6 08 46 PM" src="https://user-images.githubusercontent.com/52937121/71937600-da82d580-317a-11ea-95fd-dcd678f21441.png">

for comparison Slack toasts are below
<img width="843" alt="Screen Shot 2020-01-07 at 6 06 00 PM" src="https://user-images.githubusercontent.com/52937121/71937946-e327db80-317b-11ea-86df-ce501f4b5a69.png">

adamjclarkson commented:
I see in the spec doc a definition for Archive Toast which I can't seem to generate in my testing. Assuming that we are not delivering that toast in this PR?

adamjclarkson commented:
> Just noting here, as this is something I observed while testing
> When both toasts (channel sidebar and new toast) are both shown side by side, the difference in design is noticeable. Is this ok?
> <img alt="Screen Shot 2020-01-07 at 6 06 21 PM" width="964" src="https://user-images.githubusercontent.com/52937121/71937595-d8207b80-317a-11ea-9349-cbdd4448995c.png">
> <img alt="Screen Shot 2020-01-07 at 6 08 46 PM" width="968" src="https://user-images.githubusercontent.com/52937121/71937600-da82d580-317a-11ea-95fd-dcd678f21441.png">
> 
> for comparison Slack toasts are below
> <img alt="Screen Shot 2020-01-07 at 6 06 00 PM" width="843" src="https://user-images.githubusercontent.com/52937121/71937946-e327db80-317b-11ea-86df-ce501f4b5a69.png">

@andrewbrown00 thoughts on this?

sudheerDev commented:
> I see in the spec doc a definition for Archive Toast which I can't seem to generate in my testing. Assuming that we are not delivering that toast in this PR?

@adamjclarkson Yes, archive toast is not part of this PR. It is needed for the new permalink view update hence it is not included in this.

adamjclarkson commented:
@Willyfrog @hmhealey @jgilliam17 @andrewbrown00 

There's some urgency attached to this one for a client. Hopefully you can review in the next couple of days. THX.

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -510,30 +713,34 @@ class PostList extends React.PureComponent {
+                                            loaderId={PostListRowListIds.OLDER_MESSAGES_LOADER}
+                                            correctScrollToBottom={this.props.atLatestPost}
+                                            onItemsRendered={this.onItemsRendered}
+                                            scrollTofailed={this.scrollTofailed}
```
Can we fix the case on this?

sudheerDev replied:
Unfortunately not. We wont be able to scroll to posts which are not mounted. It is something that is really hard to do so it easier to understand that a post is not mounted and use the index to try and figure out to position the UI

hmhealey replied:
I mean, can we fix the case of the name? It should be `scrollToFailed` instead of `scrollTofailed`

sudheerDev replied:
Oops sorry got it.

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -443,6 +577,88 @@ class PostList extends React.PureComponent {
+            msgSince = (
+                <FormattedMessage
+                    id='postlist.toast.since'
+                    defaultMessage=' since {date} at {time}'
```
I don't think we can combine these two sentences in all languages. Instead, we could have two strings, one for desktop and one for mobile and switch between them as necessary.

sudheerDev replied:
Aah sorry, forgot about this. WIll update this

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -354,19 +435,56 @@ class PostList extends React.PureComponent {
+                });
+                return;
+            }
+            this.hideNewMessagesTimeout = setTimeout(() => {
```
Could the Toast component itself be in charge of fading itself out? The react-bootstrap components do that by having an internal `show` state that is changed based on the `show` prop provided by the caller.

Willyfrog replied:
also, if there needs to be a different behaviour depending on which toast it is, you could probably pass that on the construction and simply having to call `hide`, letting the component handle that.

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -118,18 +132,20 @@ class PostList extends React.PureComponent {
 
         this.listRef = React.createRef();
         this.postListRef = React.createRef();
+        this.forceUnreadEvenAtBottom = true;
```
What does this do?

Willyfrog replied:
I believe this is to prevent the toast from fading when there is a timeout sequence.

sudheerDev replied:
Normally toasts disappear when at bottom quickly but if when landing in on a channel with unreads and is at bottom(less than a page of unreads will keep the scroll position to bottom). This forces the toast to stay for a little while.

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -118,18 +132,20 @@ class PostList extends React.PureComponent {
             isScrolling: false,
             isMobile,
-            atBottom: false,
+            atBottom: true,
```
If this always starts as true, does that mean the toast will always appear when this component is first mounted?

Willyfrog replied:
it's been a long time, but i believe this was because of how it behave when mounting, sudheer might confirm this

sudheerDev replied:
Kinda of opposite, It starts with `true` now so the toast is not shown on mount if it is not needed. Otherwise i have seen flickers of this at times

sudheerDev replied:
Reverted this as i ended up changing the way things are structured.

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -178,32 +198,49 @@ class PostList extends React.PureComponent {
+        const postsAddedAtBottom = presentPostsCount !== prevPostsCount && this.props.postListIds[0] !== prevProps.postListIds[0];
+        const notBottomWithLatestPosts = !this.state.atBottom && this.props.atLatestPost && presentPostsCount > 0;
+
+        //Marking exiting messages as read based on last time user reached to the bottom
```
```suggestion
        //Marking existing messages as read based on last time user reached to the bottom
```

hmhealey commented:
For the second review assigned to @Willyfrog, we might need to get someone else to do that review since I don't know when he'll be back, and even when he does get back, he'll likely have to spend some time catching up on things.

Willyfrog commented:
I'm back, but I've been with a headache the whole week so I've been postponing the review to do a proper job. I'll try to have the review tonight

Willyfrog commented on components/post_view/post_list_virtualized/index.jsx:
```diff
@@ -33,21 +33,37 @@ export function makeCountUnreadsBelow() {
+            if (postIds) {
+                postIds = preparePostIdsForPostList(state, {postIds, lastViewedAt, channelId: ownProps.channelId});
+            }
+            newRecentMessagesCount = countUnreadsBelow(state, postIds, lastViewedAt);
```
if `postIds` is falsy, how will behave the selector? if we can ensure is an empty array, then it's fine, but there are other falsy values, maybe we should act a little defensive here.
```suggestion
            newRecentMessagesCount = countUnreadsBelow(state, (postIds || []), lastViewedAt);
```
or maybe, knowing that it is empty, have a function that returns earlier?

sudheerDev replied:
Selector has a check for 

```
            if (!postIds) {
                return 0;
            }
```

Wouldn't that suffice?

Willyfrog commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -232,6 +293,16 @@ class PostList extends React.PureComponent {
 
+    handleShortcut = (e) => {
+        if (Utils.isKeyPressed(e, Constants.KeyCodes.ESCAPE)) {
+            if (this.state.showUnreadToast) {
```
can both toasts be present at any given time? if so, shoudl we dismiss both or just one? is this the right order? I'm guessing that in case I can have both, the new messages one should be on top and I would expect to close that first, but I'd have to test that to be 100% sure

sudheerDev replied:
Both toasts will not be present at the same time. Only one exists so i don't think the order matters here.

Willyfrog commented on components/toast/toast.jsx:
```diff
@@ -0,0 +1,103 @@
+        return (
+            <div
+                className={toastClass}
+                style={{zIndex: MIN_TOAST_ZINDEX}}
```
I don't think you are using the z-ordering anymore, right? if so, we should remove it, originally it was dynamic and that was the reason for having it inline, if there is any need and it's static we should move it to a class.

sudheerDev replied:
Yup, removing this and adding it to the class

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

sudheerDev commented:
@hmhealey Thanks. For the review will get back to your review comments today.

Willyfrog commented on components/toast/toast.jsx:
```diff
@@ -0,0 +1,103 @@
+        onDismiss: PropTypes.func,
+        children: PropTypes.element,
+        show: PropTypes.bool.isRequired,
+        showActions: PropTypes.bool, //used for showing jump actions
```
can't we infer this from `onClickMessage` not being empty? or `onClick` not being null (although i can see someone passing a noop function, so maybe the message is better.

sudheerDev replied:
This is used for hiding the actions of `jump` when at bottom and the toast is present(can be for few secs)

mattermod commented:
Mattermost test server updated with git commit `6595c873bed9e64b441d1f701dc6af23a42d152b`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

andrewbrown00 commented:
> > Just noting here, as this is something I observed while testing
> > When both toasts (channel sidebar and new toast) are both shown side by side, the difference in design is noticeable. Is this ok?
> > <img alt="Screen Shot 2020-01-07 at 6 06 21 PM" width="964" src="https://user-images.githubusercontent.com/52937121/71937595-d8207b80-317a-11ea-9349-cbdd4448995c.png">
> > <img alt="Screen Shot 2020-01-07 at 6 08 46 PM" width="968" src="https://user-images.githubusercontent.com/52937121/71937600-da82d580-317a-11ea-95fd-dcd678f21441.png">
> > for comparison Slack toasts are below
> > <img alt="Screen Shot 2020-01-07 at 6 06 00 PM" width="843" src="https://user-images.githubusercontent.com/52937121/71937946-e327db80-317b-11ea-86df-ce501f4b5a69.png">
> 
> @andrewbrown00 thoughts on this?

@adamjclarkson it's okay for now, we will add the consistency as part of the UI Redesign effort which is in-progress. If we have time we could create a ticket to look at those banner buttons for the next release and get them updated globally.

sudheerDev commented on package.json:
```diff
@@ -56,7 +56,7 @@
     "react-transition-group": "4.3.0",
     "react-virtualized-auto-sizer": "1.0.2",
-    "react-window": "github:mattermost/react-window#fcc3f86581fc0bb4a1f34c0024b34f06fc9d971a",
+    "react-window": "github:mattermost/react-window#b5682113b2f5ad68ffbe596e99e60a3903f91776",
```
Ya will raise a PR for it. needed a minor modification to the library

sudheerDev replied:
https://github.com/mattermost/react-window/pull/25

hmhealey replied:
Also, this hash needs to be updated once the other PR is merged

sudheerDev commented:
> I'm a bit worried how much logic that the PostListVirtualized needs to control the Toast. PostListVirtualized is already rather bloated, and this is adding more hard-to-follow logic to it.

@hmhealey I agree, When i started with this i didn't realise it will complicate this. Looking at making a wrapper and separate some logic from here

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `50c6d4d3d8d11b97ca2babcba9b72d9bd3de173d`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

sudheerDev commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -480,6 +683,7 @@ class PostList extends React.PureComponent {
                             isScrolling={this.state.isScrolling}
                             isMobile={true}
                             postId={this.state.topPostId}
+                            stylesOverride={this.state.showUnreadToast || this.state.showNewMessagesToast ? {top: '50px'} : null}
```
It is too specific and i don't see a really good value of adding a class for this. It is hard to maintain if it is anyway not reusable.

sudheerDev commented:
@hmhealey I changed the way toasts is structured to remove some of the complexity from virt list. I also changed the styles to be in toast.scss along with internal state handling timeouts

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `df86f547189242a25d3b57da30c97460b8bad3b6`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

sudheerDev commented on components/user_settings/notifications/email_notification_setting/__snapshots__/email_notification_setting.test.jsx.snap:
```diff
@@ -334,8 +334,8 @@ exports[`components/user_settings/notifications/EmailNotificationSetting should
 <Connect(SettingItemMin)
   describe={
     <FormattedMessage
-      defaultMessage="Immediately"
```
i don't know why this was falling for me

sudheerDev replied:
Never mind this. Updating mm-redux somehow fixed this.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

sudheerDev commented on components/toast/toast.jsx:
```diff
@@ -0,0 +1,139 @@
+
+    componentWillUnmount() {
+        this.mounted = false;
+        clearTimeout(this.toastTimer);
```
Nope not a problem even if it is already ran or never existed even

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Mattermost test server updated with git commit `af652b7500343284a150a6345b6c495edbeb35a2`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `942a066d119120eb1843b94688a034a8bc119916`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `4aa1ce5b1cbc20ea0141b88a7928814456f6f84b`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

sudheerDev commented:
@andrewbrown00 @hmhealey @Willyfrog Up for review again.

hmhealey commented on components/toast_wrapper/toast_wrapper.jsx:
```diff
@@ -0,0 +1,253 @@
+        }
+    }
+
+    newMessagesToastText = (count, since) => {
```
This still needs to be fixed. We should just have two messages and alternate between since combining strings doesn't work the same in all languages.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `26faef22d7bc2a0f18a786072ff151faa9217c71`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `0a30b12ada72bcac2e44ff5f639827c7b8f47b86`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `5fa02bd9994ccccd3e92187fd7de6e8bd7ebe43f`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `1804eb90eff741ccb343d33867289121b71e6d1c`.

Access here: https://mattermost-webapp-pr-4628.test.mattermost.cloud

mattermod commented:
Test server destroyed
